flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -1362,16 +1356,12 @@ private boolean transitionState(JobStatus current, JobStatus newState, Throwable
 		}
 
 		// now do the actual state transition
-		if (STATE_UPDATER.compareAndSet(this, current, newState)) {
-			LOG.info("Job {} ({}) switched from state {} to {}.", getJobName(), getJobID(), current, newState, error);
+		state = newState;
+		LOG.info("Job {} ({}) switched from state {} to {}.", getJobName(), getJobID(), current, newState, error);
 
-			stateTimestamps[newState.ordinal()] = System.currentTimeMillis();
-			notifyJobStatusChange(newState, error);
-			return true;
-		}
-		else {
-			return false;
-		}
+		stateTimestamps[newState.ordinal()] = System.currentTimeMillis();
+		notifyJobStatusChange(newState, error);
+		return true;
```

Comment:
We cannot always set `newState` and return `true`. 

`current` is meant to be the expected current state. There are plenty of places where we throw `IllegalStateException` if the expected state does not match. For example https://github.com/apache/flink/blob/b7cd6a984750d7dd98e6408f3703b1978b14355a/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java#L985-L986

Reply:
You are right. I have updated the PR. What do you think about the new change?